src/server/src/main/java/com/spotify/reaper/service/SegmentRunner.java

```diff
@@ -570,32 +570,53 @@ public void handle(int repairNumber, Optional<ActiveRepairService.Status> status
     synchronized (condition) {
       RepairSegment currentSegment = context.storage.getRepairSegment(repairRunner.getRepairRunId(), segmentId).get();
       // See status explanations at: https://wiki.apache.org/cassandra/RepairAsyncAPI
-      // Old repair API
+      // Old repair API – up to Cassandra-2.1.x
       if(status.isPresent()) {
         switch (status.get()) {
           case STARTED:
-            renewLead();
-            DateTime now = DateTime.now();
-            context.storage.updateRepairSegment(currentSegment.with()
-                .state(RepairSegment.State.RUNNING)
-                .startTime(now)
-                .build(segmentId));
-            LOG.debug("updated segment {} with state {}", segmentId, RepairSegment.State.RUNNING);
+            try {
+              if (renewLead()) {
+                DateTime now = DateTime.now();
+                
+                context.storage.updateRepairSegment(currentSegment.with()
+                    .state(RepairSegment.State.RUNNING)
+                    .startTime(now)
+                    .build(segmentId));
+                
+                LOG.debug("updated segment {} with state {}", segmentId, RepairSegment.State.RUNNING);
+                break;
+              }
+            } catch (AssertionError er) {
+              // ignore. segment repair has since timed out.
+            }
+            segmentFailed.set(true);
             break;
 
           case SESSION_SUCCESS:
-            if (timedOut.get()) {
-              LOG.debug("Got SESSION_SUCCESS for segment with id '{}' and repair number '{}', " +
-                        "but it had already timed out",
-                        segmentId, repairNumber);
-            } else {
-              LOG.debug("repair session succeeded for segment with id '{}' and repair number '{}'",
-                      segmentId, repairNumber);
-              context.storage.updateRepairSegment(currentSegment.with()
-                  .state(RepairSegment.State.DONE)
-                  .endTime(DateTime.now())
-                  .build(segmentId));
+            try {
+              if (segmentFailed.get()) {
+                LOG.debug(
+                    "Got SESSION_SUCCESS for segment with id '{}' and repair number '{}', but it had already timed out",
+                    segmentId, repairNumber);
+              } else if (renewLead()) {
+                try {
+                  LOG.debug("repair session succeeded for segment with id '{}' and repair number '{}'", segmentId,
+                      repairNumber);
+
+                  context.storage.updateRepairSegment(currentSegment.with()
+                      .state(RepairSegment.State.DONE)
+                      .endTime(DateTime.now())
+                      .build(segmentId));
+
+                  break;
+                }  finally {
+                  releaseLead();
```

Comment:
It seems like we shouldn't release the lead here. 
What happens in 2.1 is that we get the following notifications in order : STARTED -> SESSION_SUCCESS -> FINISHED.
`FINISHED` is the one that will trigger the release of the condition, which will make us exit the loop in `runRepair()` and subsequently call `renewLead()` (which fails here because we released it already). 
Then, we release the lead whatever happens at the end of `run()` in the `finally` block.
I'd change the code here by removing the `finally` block and instead silently ignore a `renewLead()` that fails.

Reply:
i presume it's the same thing with 2.2+
`START -> SUCCESS -> COMPLETE` ?

Reply:
Presuming code, have fixed the code in both places. If you could test again @adejanovski that would be great.